File: es/01-programming/javascript/ejercicios/nivel-05-experto/ejercicio-02-servidor-http.test.js
const { test, before, after } = require("node:test");
const assert = require("node:assert/strict");
const { crearServidor } = require("./ejercicio-02-servidor-http");

let server;
let base;

before(async () => {
  server = crearServidor();
  await new Promise((resolve) => server.listen(0, resolve));
  base = `http://localhost:${server.address().port}`;
});

after(() => new Promise((resolve) => server.close(resolve)));

test("GET / responde con la bienvenida", async () => {
  const res = await fetch(`${base}/`);
  assert.equal(res.status, 200);
  assert.equal(await res.text(), "Bienvenido al servidor de Node.js");
});

test("GET /hora devuelve una fecha ISO válida", async () => {
  const res = await fetch(`${base}/hora`);
  assert.equal(res.status, 200);
  const cuerpo = await res.text();
  assert.ok(!Number.isNaN(Date.parse(cuerpo)));
});

test("GET /saludo?nombre=Ana saluda a Ana", async () => {
  const res = await fetch(`${base}/saludo?nombre=Ana`);
  assert.equal(res.status, 200);
  assert.equal(await res.text(), "Hola, Ana!");
});

test("GET /saludo sin nombre usa 'mundo'", async () => {
  const res = await fetch(`${base}/saludo`);
  assert.equal(await res.text(), "Hola, mundo!");
});

test("GET /productos devuelve el JSON de productos", async () => {
  const res = await fetch(`${base}/productos`);
  assert.equal(res.status, 200);
  const datos = await res.json();
  assert.equal(datos.length, 2);
  assert.equal(datos[0].nombre, "Laptop");
});

test("ruta desconocida devuelve 404", async () => {
  const res = await fetch(`${base}/otra`);
  assert.equal(res.status, 404);
  assert.equal(await res.text(), "No encontrado");
});

File: es/01-programming/csharp/ejercicios/nivel-05-experto/ejercicio-02-servidor-http_test.cs
using System;

public static class Programa
{
    private static int _fallos;

    private static void Check(string nombre, Func<bool> prueba)
    {
        try
        {
            if (prueba())
            {
                Console.WriteLine("[OK]   " + nombre);
            }
            else
            {
                Console.WriteLine("[FALL] " + nombre);
                _fallos++;
            }
        }
        catch (Exception ex)
        {
            Console.WriteLine("[FALL] " + nombre + " -> " + ex.GetType().Name + ": " + ex.Message);
            _fallos++;
        }
    }

    public static int Main()
    {
        ServidorHttp servidor = Ejercicio02.CrearServidor(8080);
        Check("CrearServidor(8080) guarda el puerto", () => servidor.Puerto == 8080);

        Check("ProcesarRuta(\"/\") devuelve '<h1>Inicio</h1>'",
            () => servidor.ProcesarRuta("/") == "<h1>Inicio</h1>");

        Check("ProcesarRuta(\"/saludo\") devuelve '<h1>Hola, bienvenido</h1>'",
            () => servidor.ProcesarRuta("/saludo") == "<h1>Hola, bienvenido</h1>");

        Check("ProcesarRuta(\"/saludo?nombre=Ana\") incluye 'Ana'",
            () => servidor.ProcesarRuta("/saludo?nombre=Ana") == "<h1>Hola, Ana</h1>");
        Check("ProcesarRuta(\"/saludo?nombre=Luis\") incluye 'Luis'",
            () => servidor.ProcesarRuta("/saludo?nombre=Luis") == "<h1>Hola, Luis</h1>");

        string hora = servidor.ProcesarRuta("/api/hora");
        Check("ProcesarRuta(\"/api/hora\") devuelve una hora HH:mm:ss",
            () => hora.Length == 8 && hora[2] == ':' && hora[5] == ':');

        Check("ProcesarRuta(\"/nope\") devuelve '404 No encontrado'",
            () => servidor.ProcesarRuta("/nope") == "404 No encontrado");
        Check("ProcesarRuta(\"\") devuelve '404 No encontrado'",
            () => servidor.ProcesarRuta("") == "404 No encontrado");

        Console.WriteLine();
        if (_fallos == 0)
        {
            Console.WriteLine("Todos los tests pasaron.");
            return 0;
        }
        Console.WriteLine(_fallos + " test(s) fallaron.");
        return 1;
    }
}